src/main/java/org/broadinstitute/consent/http/service/DatasetService.java

```diff
@@ -57,6 +57,36 @@ public DataSet getDatasetWithPropertiesById(Integer datasetId) {
         return dataset;
     }
 
+    public DataSet updateDataset(DataSetDTO dataset, Integer datasetId, Integer userId) {
+        Timestamp now = new Timestamp(new Date().getTime());
+
+        DataSet old = getDatasetWithPropertiesById(datasetId);
+        Set<DataSetProperty> oldProperties = old.getProperties();
+
+        List<DataSetPropertyDTO> dtos = dataset.getProperties();
+        List<DataSetProperty> jsonProperties = processDatasetProperties(datasetId, dtos);
+              ;
+        List<DataSetProperty> propertiesToAdd = jsonProperties.stream()
+              .filter(p -> oldProperties.stream()
+                    .noneMatch(op -> op.getPropertyKey() == p.getPropertyKey()))
+              .collect(Collectors.toList());
+
+        List<DataSetProperty> propertiesToUpdate = jsonProperties.stream()
+              .filter(p -> oldProperties.stream()
+                    .noneMatch(op -> equalsProperty(op.getPropertyKey(), op.getPropertyValue(), p.getPropertyKey(), p.getPropertyValue())))
+              .collect(Collectors.toList());
+
+        if (propertiesToAdd.isEmpty() && propertiesToUpdate.isEmpty()) {
+            return null;
```

Comment:
Couldn't think of a better way to distinguish "not modified" datasets, but I wanted to make sure this was checked for so we don't update the dataset (with date, user id) if no properties were actually updated.

Reply:
A clearer way to handle this is to return and `Optional<Dataset>` instead of just a `Dataset`. That way, callers know that the dataset is not guaranteed and don't have to do null checking. If you just return Optional.empty here, the caller would know nothing happened.